File: src/osprey/interfaces/artifacts/static/js/timeseries.js
// @ts-check
/**
 * OSPREY Artifact Gallery — timeseries preview: the lazy Plotly loader,
 * the toolbar/chart/table viewer, and CSV/JSON export.
 *
 * `renderTimeseriesView(container, artifact)` is the module's single entry
 * point, called by preview.js's `renderPreview` via the `onTimeseriesNeeded`
 * callback, which gallery.js wires to the export here.
 *
 * No factory/closure state is needed: `_plotlyLoaded`/`_plotlyLoading` are a
 * true page-lifetime singleton cache (matches types.js's `typeRegistry`
 * pattern — there is only ever one gallery per page, and Plotly only ever
 * needs to load once), and the channel-visibility `Set` in
 * `renderTimeseriesView` is local to a single render call, not shared state
 * across calls. So this module is plain exports, no `createXRenderer(...)`.
 *
 * The lazy `<script src="/static/js/vendor/plotly-3.3.1.min.js">` injection
 * must keep that exact offline-vendored path — do not change it.
 *
 * HTML-escaping uses the design-system's canonical `escapeHtml` (quote-safe,
 * nullish-collapsing) — see dom.js. This module used to carry its own
 * `_esc` copy; that divergence has been consolidated away.
 *
 * Table cells apply magnitude-adaptive formatting: index cells go through
 * `_tsShortTime` (short month/day + hour:minute:second, no year, since the
 * backend's `index` is ISO timestamp strings — see
 * src/osprey/utils/timeseries.py's `extract_channel_series`), value cells
 * through `_tsFormatValue` (<=5 significant figures, scientific notation for
 * very large/small magnitudes). Both helpers fall back to raw `String(...)`
 * for inputs that aren't a number/valid date, so their output is still run
 * through `escapeHtml` at the interpolation site — never trust it as safe
 * HTML on its own.
 *
 * @module timeseries
 */

import { chartTheme, chartSeries } from "/design-system/js/theme-manager.js";
import { escapeHtml } from "/design-system/js/dom.js";
import { isoToDate } from "./types.js";

// ---- Lazy Plotly Loader ---- //

let _plotlyLoaded = false;
/** @type {Promise<void>|null} */
let _plotlyLoading = null;

/** @returns {Promise<void>} */
function ensurePlotlyLoaded() {
  if (_plotlyLoaded) return Promise.resolve();
  if (_plotlyLoading) return _plotlyLoading;
  _plotlyLoading = new Promise((resolve, reject) => {
    const script = document.createElement("script");
    script.src = "/static/js/vendor/plotly-3.3.1.min.js";
    script.onload = () => { _plotlyLoaded = true; resolve(); };
    script.onerror = () => {
      _plotlyLoading = null;
      reject(new Error("Failed to load Plotly"));
    };
    document.head.appendChild(script);
  });
  return _plotlyLoading;
}

// ---- Helpers ---- //

/**
 * @param {string} name
 * @returns {string}
 */
function _tsShortChannelName(name) {
  if (!name) return "";
  if (name.length <= 24) return name;
  const parts = name.split(":");
  if (parts.length >= 3) return parts[0] + ":...:" + parts[parts.length - 1];
  return name.slice(0, 10) + "..." + name.slice(-10);
}

/**
 * Magnitude-adaptive value-cell formatter: <=5 significant figures for
 * ordinary magnitudes, scientific notation once a value is very large or
 * very small (but nonzero). Falls back to `String(...)` for non-number
 * (including NaN) input — callers MUST still escapeHtml the result, since
 * that fallback echoes the raw input verbatim.
 * @param {any} num
 * @returns {string}
 */
function _tsFormatValue(num) {
  if (num === null || num === undefined) return "--";
  if (typeof num !== "number" || Number.isNaN(num)) return String(num);
  if (num === 0) return "0";
  const abs = Math.abs(num);
  if (abs >= 1e6 || abs < 0.001) return num.toExponential(3);
  return num.toPrecision(5);
}

/**
 * Short index/time-cell formatter for ISO timestamp strings: month/day +
 * hour:minute:second, no year. Shares types.js's `isoToDate` guard, which
 * rejects the null/number/numeric-string inputs that bare `Date` coercion
 * would otherwise turn into fabricated epoch/year-2000 timestamps: nullish
 * input renders "--", and any other non-ISO/invalid value falls back to
 * `String(iso)` verbatim. Callers MUST still escapeHtml the result, since
 * that fallback echoes the raw input.
 * @param {any} iso
 * @returns {string}
 */
function _tsShortTime(iso) {
  if (iso === null || iso === undefined) return "--";
  const d = isoToDate(iso);
  if (!d) return String(iso);
  return d.toLocaleString(undefined, {
    month: "short",
    day: "numeric",
    hour: "2-digit",
    minute: "2-digit",
    second: "2-digit",
  });
}

/**
 * Whether one channel is enum/status rather than a numeric signal.
 *
 * Deliberately tri-state: ONLY an explicit `numeric === false` means enum, so a
 * response that omits the key is still numeric. Four sites ask this question --
 * whether the layout needs a `yaxis2`, whether the toolbar marks the toggle,
 * whether hiding a channel hides that axis, and which axis a trace is built on
 * -- and they must not drift, since getting the last one wrong routes a trace
 * onto the wrong axis rather than failing loudly.
 * @param {any} ch
 * @returns {boolean}
 */
function _tsIsNonNumeric(ch) {
  return ch.numeric === false;
}

/**
 * Whether any channel is enum/status, i.e. whether the chart needs the
 * secondary category `yaxis2`.
 *
 * Both the view shell and the chart renderer need this answer and must agree:
 * the shell uses it to autorange and show/hide that axis on Reset Zoom and on
 * channel toggles, and the renderer uses it to decide whether to put the axis
 * in the layout at all. Computed once here so the two cannot drift.
 * @param {any[]} channels
 * @returns {boolean}
 */
function _tsHasNonNumeric(channels) {
  return (channels || []).some(_tsIsNonNumeric);
}

/**
 * RFC4180-ish CSV field quoting: a field containing a comma, double quote, or
 * line break is wrapped in double quotes with internal double quotes doubled.
 * Needed now that a CSV row's fields include free-form channel names and
 * non-numeric (enum/status) values -- e.g. a status string like
 * `"OFF, LOCAL"` -- rather than only numbers, which never need escaping.
 * @param {any} value
 * @returns {string}
 */
function _tsCsvField(value) {
  // `??`, not `||`: a gap (null/undefined) becomes an empty field, but the
  // falsy-but-real values `0` and `""` must survive as themselves.
  const str = String(value ?? "");
  if (/["\n\r,]/.test(str)) {
    return `"${str.replace(/"/g, '""')}"`;
  }
  return str;
}

/**
 * Long-format CSV: one row per (channel, timestamp) pair. Each channel now
 * carries its own timestamps independent of every other channel's (see
 * app.py's `format=chart` branch), so there is no shared axis to pivot into
 * a wide `timestamp,ch1,ch2,...` table without re-implementing the
 * server-side union pivot client-side (that pivot already exists, for the
 * table view, via `format=table`). Long format is the honest shape of the
 * data actually on hand here.
 * @param {any} chartData
 */
function _tsExportCSV(chartData) {
  const channels = chartData.channels || [];
  const rows = channels.flatMap((/** @type {any} */ ch) =>
    (ch.timestamps || []).map((/** @type {any} */ ts, /** @type {number} */ i) =>
      [ch.channel, ts, (ch.values || [])[i]].map(_tsCsvField).join(",")
    )
  );
  _tsDownloadBlob(["channel,timestamp,value", ...rows].join("\n"), "text/csv", "csv");
}

/** @param {any} chartData */
function _tsExportJSON(chartData) {
  _tsDownloadBlob(JSON.stringify(chartData, null, 2), "application/json", "json");
}

/**
 * Push `content` at the browser as a download. The two exporters differ only
 * in their content, MIME type and extension, so the Blob/ObjectURL/anchor
 * sequence -- including the revoke, which is easy to drop when copied -- lives
 * here once.
 * @param {string} content
 * @param {string} mime
 * @param {string} ext
 */
function _tsDownloadBlob(content, mime, ext) {
  const url = URL.createObjectURL(new Blob([content], { type: mime }));
  const a = document.createElement("a");
  a.href = url;
  a.download = `timeseries_${Date.now()}.${ext}`;
  a.click();
  URL.revokeObjectURL(url);
}

// ---- Timeseries View (toolbar + chart + table) ---- //

/**
 * @param {HTMLElement} container
 * @param {any} artifact
 * @returns {Promise<void>}
 */
export async function renderTimeseriesView(container, artifact) {
  container.innerHTML = '<div class="ts-loading">Loading timeseries data...</div>';
  try {
    const chartResp = await fetch(
      `/api/artifacts/${artifact.id}/data?format=chart&max_points=2000`
    );
    if (!chartResp.ok) throw new Error(`Chart fetch failed: ${chartResp.status}`);
    const chartData = await chartResp.json();
    // format=chart returns one entry per channel, each with its own
    // timestamps/values (see app.py's chart branch) -- there is no shared
    // `columns`/`total_rows` frame any more. `columns` here is just the
    // channel-name projection, kept for the toggle logic and for
    // renderTimeseriesTable, which still consumes the (unchanged) `format=table`
    // shape and expects a plain name list.
    const channels = chartData.channels || [];
    const columns = channels.map((/** @type {any} */ ch) => ch.channel);
    // Whether renderTimeseriesChart will add a secondary category `yaxis2`
    // for any enum/status channel -- needed here too, both so Reset Zoom
    // also autoranges that axis (Plotly.relayout's autorange keys are
    // per-axis; `yaxis.autorange` alone never touches `yaxis2`) and so a
    // channel toggle can show/hide that axis as its last visible occupant
    // is hidden/shown (see the toggle handler below).
    const hasNonNumeric = _tsHasNonNumeric(channels);

    const visible = new Set(columns);

    let html = '<div class="ts-viewer">';

    // Info bar
    html += '<div class="ts-info-bar">';
    channels.forEach((/** @type {any} */ ch) => {
      html += `<span class="ts-badge ts-badge-channel"><span class="badge-label">CH</span> ${escapeHtml(_tsShortChannelName(ch.channel))}</span>`;
    });
    // Cross-channel totals are the server's, unconditionally. Summing per-channel
    // point counts here could only ever be a proxy: the sum counts each channel's
    // own timestamps, whereas the table paginates a *unioned* row axis, so the two
    // disagree whenever channels are sampled at different times -- and `row_count`
    // is not derivable client-side at all. This script and the endpoint that
    // answers it ship in the same wheel, so `summary` is always present; a
    // fallback here would be a second, knowingly-wrong number for a case that
    // cannot arise.
    const summary = chartData.summary;
    html += `<span class="ts-badge ts-badge-points"><span class="badge-label">Points</span> ${summary.total_points.toLocaleString()}</span>`;
    if (typeof summary.row_count === "number") {
      html += `<span class="ts-badge ts-badge-rows"><span class="badge-label">Rows</span> ${summary.row_count.toLocaleString()}</span>`;
    }
    if (summary.downsampled) {
      html += `<span class="ts-badge ts-badge-downsampled"><span class="badge-label">Downsampled</span> ${summary.returned_points.toLocaleString()} pts</span>`;
    }
    html += '</div>';

    // Toolbar
    html += '<div class="ts-toolbar">';
    html += '<div class="ts-toolbar-group">';
    const _tsPalette = chartSeries();
    channels.forEach((/** @type {any} */ ch, /** @type {number} */ ci) => {
      const color = _tsPalette[ci % _tsPalette.length];
      const isNonNumeric = _tsIsNonNumeric(ch);
      const title = isNonNumeric ? `${ch.channel} (status/enum -- plotted on right axis)` : ch.channel;
      // `aria-pressed` + an explicit `aria-label` mirror the design system's own
      // toggle (osprey-theme-switcher): the button's shown/hidden state must be
      // exposed to assistive tech rather than carried only by the `ts-ch-off`
      // class, and the accessible name must be the full PV plus the axis
      // explanation -- not the truncated display text with the decorative "R"
      // glyph welded on (accname demotes `title` to a description whenever the
      // button has content, so `title` alone often goes unannounced).
      html += `<button class="ts-ch-toggle" type="button" aria-pressed="true" data-ch-name="${escapeHtml(ch.channel)}" aria-label="${escapeHtml(title)}" title="${escapeHtml(title)}">`;
      html += `<span class="ts-ch-dot" style="background:${color}"></span>`;
      html += escapeHtml(_tsShortChannelName(ch.channel));
      if (isNonNumeric) html += ' <span class="ts-ch-axis-tag" aria-hidden="true">R</span>';
      html += '</button>';
    });
    html += '</div>';
    html += '<span class="ts-toolbar-divider"></span>';
    html += '<div class="ts-toolbar-group">';
    html += '<button class="ts-action-btn" data-action="zoom-reset" title="Reset zoom">Reset Zoom</button>';
    html += '<button class="ts-action-btn ts-export-btn" data-action="export-csv" title="Export CSV">CSV</button>';
    html += '<button class="ts-action-btn ts-export-btn" data-action="export-json" title="Export JSON">JSON</button>';
    html += '</div>';
    html += '</div>';

    // Chart
    html += '<div class="ts-chart-container" data-ts-chart></div>';

    // Table
    html += '<div data-ts-table></div>';
    html += '</div>';
    container.innerHTML = html;

    // Wire events
    const chartEl = container.querySelector("[data-ts-chart]");

    container.querySelectorAll(".ts-ch-toggle").forEach((btn) => {
      btn.addEventListener("click", () => {
        const col = /** @type {string} */ (/** @type {HTMLElement} */ (btn).dataset.chName);
        if (visible.has(col)) {
          if (visible.size <= 1) return;
          visible.delete(col);
          btn.classList.add("ts-ch-off");
        } else {
          visible.add(col);
          btn.classList.remove("ts-ch-off");
        }
        // Keep the exposed state in step with the class (the refused-hide path
        // above returns early, so neither is touched there).
        btn.setAttribute("aria-pressed", String(visible.has(col)));
        if (chartEl && /** @type {any} */ (chartEl).data) {
          // `columns` is in the same order as `channels`, which is the same
          // order traces were built in (renderTimeseriesChart), so this
          // boolean array lines up positionally with Plotly's trace array
          // even though visibility itself is tracked by channel name.
          const update = columns.map((/** @type {any} */ c) => visible.has(c));
          // Hiding the last visible enum/status channel would otherwise leave an
          // empty category `yaxis2` sitting on the right with nothing plotted
          // against it; show it again once any non-numeric channel is visible
          // again. Applied as one `update` rather than restyle-then-relayout so
          // a toggle costs a single redraw instead of two, with no intermediate
          // frame showing the traces already hidden but the axis still there.
          const layoutUpdate = hasNonNumeric
            ? {
                "yaxis2.visible": channels.some(
                  (/** @type {any} */ ch) => _tsIsNonNumeric(ch) && visible.has(ch.channel)
                ),
              }
            : {};
          Plotly.update(chartEl, { visible: update }, layoutUpdate);
        }
      });
    });

    container.querySelectorAll(".ts-action-btn").forEach((btn) => {
      btn.addEventListener("click", () => {
        const action = /** @type {HTMLElement} */ (btn).dataset.action;
        if (action === "zoom-reset" && chartEl) {
          // `yaxis2` (the enum/status channels' secondary category axis, see
          // renderTimeseriesChart) predates this handler and needs its own
          // autorange key -- otherwise a user who zooms it and clicks Reset
          // Zoom gets the numeric axis reset while the status trace stays
          // clipped or off-screen, reading as "that channel has no data".
          const resetLayout = {
            "xaxis.autorange": true,
            "yaxis.autorange": true,
            ...(hasNonNumeric ? { "yaxis2.autorange": true } : {}),
          };
          Plotly.relayout(chartEl, resetLayout);
        } else if (action === "export-csv") {
          _tsExportCSV(chartData);
        } else if (action === "export-json") {
          _tsExportJSON(chartData);
        }
      });
    });

    await renderTimeseriesChart(/** @type {any} */ (chartEl), chartData);

    const tableEl = container.querySelector("[data-ts-table]");
    await renderTimeseriesTable(/** @type {any} */ (tableEl), artifact.id, 0);
  } catch (err) {
    console.error("Timeseries render failed:", err);
    container.innerHTML = '<span class="text-muted">Failed to load timeseries data</span>';
  }
}

/**
 * A Plotly layout fragment for the timeseries chart, built from
 * chartTheme()'s --chart-* computed-style bridge plus a couple of
 * gallery-specific extras (axis/legend line color, legend background)
 * that bridge doesn't cover -- read directly via getComputedStyle so
 * they still track the design tokens rather than being hardcoded.
 * @returns {any}
 */
export function _tsChartTheme() {
  const base = chartTheme();
  const line = getComputedStyle(document.documentElement).getPropertyValue("--border-default").trim();
  return { ...base, line: line || base.xaxis.gridcolor, legendBg: base.paper_bgcolor, legendBorder: line };
}

/**
 * Builds one Plotly trace per channel, each with its own `x`/`y` (Plotly
 * supports per-trace `x` natively -- there is no shared axis to slice
 * multiple columns against any more, see app.py's `format=chart` branch and
 * `extract_channel_series`'s docstring).
 *
 * A non-numeric (enum/status) channel -- `numeric: false` -- can't share a
 * linear y-axis with a numeric signal like beam current, so it is never
 * silently dropped: it gets its own categorical trace on a secondary
 * right-hand y-axis (`yaxis2`, `type: "category"`, `overlaying: "y"`), drawn
 * with an `hv` step line (a status value holds until its next transition,
 * unlike a numeric signal that's meaningfully interpolated between samples).
 * This lets an operator read machine state time-aligned against a numeric
 * signal on the same plot, rather than losing the channel or corrupting the
 * numeric axis. The secondary axis is only added to the layout when at
 * least one channel actually needs it.
 * @param {any} el
 * @param {any} chartData
 * @returns {Promise<void>}
 */
export async function renderTimeseriesChart(el, chartData) {
  await ensurePlotlyLoaded();
  if (!el) return;

  const channels = chartData.channels || [];
  const hasNonNumeric = _tsHasNonNumeric(channels);

  const traces = channels.map((/** @type {any} */ ch) => {
    const numeric = !_tsIsNonNumeric(ch);
    return {
      x: ch.timestamps,
      y: ch.values,
      name: ch.channel,
      // scattergl (WebGL) is faster for the common numeric case; a
      // categorical y-axis is safer on the plain SVG "scatter" renderer.
      type: numeric ? "scattergl" : "scatter",
      mode: numeric ? "lines" : "lines+markers",
      ...(numeric
        ? {}
        : {
            yaxis: "y2",
            line: { shape: "hv" },
            // Every enum channel shares one category axis, and Plotly builds
            // that axis's rungs from the union of its traces in first-
            // appearance order. Two status channels with different vocabularies
            // therefore interleave: each one's step line crosses rungs that
            // belong to the other, which reads as a state it was never in.
            // Prefixing each value with its channel keeps every channel's rungs
            // in its own contiguous block. `customdata` carries the unprefixed
            // value so the hover label still shows the real state.
            //
            // A gap (`null` -- archiver_read.py's chart branch emits real
            // `None`s for missing samples, and extract_channel_series passes
            // them through) is deliberately NOT namespaced: prefixing it would
            // register the literal rung "CH: null" on the category axis and
            // step the line up to it, i.e. show a state the channel was never
            // in -- the very failure this namespacing exists to prevent. Left
            // as `null`, Plotly's category converter returns BADNUM instead of
            // adding a category, and the step line breaks, matching how the
            // numeric branch already renders the same gap.
            y: (ch.values || []).map((/** @type {any} */ v) => (v == null ? null : `${ch.channel}: ${v}`)),
            customdata: ch.values,
          }),
      hovertemplate: numeric
        ? "%{y:.4g}<extra>%{fullData.name}</extra>"
        : "%{customdata}<extra>%{fullData.name}</extra>",
    };
  });

  const t = _tsChartTheme();

  const layout = {
    paper_bgcolor: t.paper_bgcolor,
    plot_bgcolor: t.plot_bgcolor,
    font: { family: "'JetBrains Mono', monospace", size: 11, color: t.font.color },
    margin: { t: 30, r: 20, b: 50, l: 60 },
    hovermode: "x unified",
    xaxis: { gridcolor: t.xaxis.gridcolor, linecolor: t.line, tickfont: { size: 10 } },
    yaxis: { gridcolor: t.yaxis.gridcolor, linecolor: t.line, tickfont: { size: 10 } },
    ...(hasNonNumeric
      ? {
          yaxis2: {
            overlaying: "y",
            side: "right",
            type: "category",
            linecolor: t.line,
            tickfont: { size: 10 },
            // This axis's tick labels are the namespaced rungs -- "<full PV
            // name>: <VALUE>" -- the longest strings in the figure, drawn on
            // the right against a `margin.r` of 20. Plotly's `automargin`
            // defaults to false for an x-anchored axis, and `margin.autoexpand`
            // only reacts to components that push margin, so without this the
            // labels are drawn past the paper edge and clipped.
            automargin: true,
            // ...and `showgrid` otherwise defaults to true with no `gridcolor`,
            // adding one gridline per rung in Plotly's auto blend rather than
            // the --chart-* token the other two axes use. The numeric `yaxis`
            // grid is the one that carries meaning here.
            showgrid: false,
          },
        }
      : {}),
    legend: { bgcolor: t.legendBg, bordercolor: t.legendBorder, borderwidth: 1, font: { size: 10 } },
    colorway: chartSeries(),
  };

  Plotly.newPlot(el, traces, layout, {
    responsive: true,
    displaylogo: false,
    modeBarButtonsToRemove: ["lasso2d", "select2d"],
  });
}

const TS_TABLE_PAGE_SIZE = 50;

/**
 * @param {any} el
 * @param {string} artifactId
 * @param {number} offset
 * @returns {Promise<void>}
 */
export async function renderTimeseriesTable(el, artifactId, offset) {
  if (!el) return;
  el.innerHTML = '<div class="ts-loading">Loading table...</div>';

  try {
    const resp = await fetch(
      `/api/artifacts/${artifactId}/data?format=table&offset=${offset}&limit=${TS_TABLE_PAGE_SIZE}`
    );
    if (!resp.ok) throw new Error(`Table fetch failed: ${resp.status}`);
    const tableData = await resp.json();

    const totalPages = Math.ceil(tableData.total_rows / TS_TABLE_PAGE_SIZE);
    const currentPage = Math.floor(offset / TS_TABLE_PAGE_SIZE) + 1;

    // Header comes from THIS response and nowhere else. Taking it from the
    // caller's channel list would mean the names came from a separate
    // format=chart request, and for an artifact being written while it is viewed
    // the two can disagree -- putting this page's cells under the other
    // request's PV names.
    let html = '<div class="ts-data-table-wrapper"><table class="ts-data-table">';
    html += '<thead><tr><th>Index</th>';
    tableData.columns.forEach((/** @type {string} */ c) => { html += `<th>${escapeHtml(c)}</th>`; });
    html += '</tr></thead><tbody>';

    tableData.index.forEach((/** @type {any} */ idx, /** @type {number} */ i) => {
      html += '<tr>';
      html += `<td class="ts-index-cell">${escapeHtml(_tsShortTime(idx))}</td>`;
      const row = tableData.data[i] || [];
      row.forEach((/** @type {any} */ val) => { html += `<td>${escapeHtml(_tsFormatValue(val))}</td>`; });
      html += '</tr>';
    });

    html += '</tbody></table></div>';

    html += '<div class="ts-pagination">';
    html += `<button class="btn btn-secondary btn-sm" data-ts-prev ${offset === 0 ? "disabled" : ""}>Prev</button>`;
    html += `<span class="ts-page-info">Page ${currentPage} of ${totalPages}</span>`;
    html += `<button class="btn btn-secondary btn-sm" data-ts-next ${offset + TS_TABLE_PAGE_SIZE >= tableData.total_rows ? "disabled" : ""}>Next</button>`;
    html += '</div>';

    el.innerHTML = html;

    el.querySelector("[data-ts-prev]")?.addEventListener("click", () => {
      renderTimeseriesTable(el, artifactId, Math.max(0, offset - TS_TABLE_PAGE_SIZE));
    });
    el.querySelector("[data-ts-next]")?.addEventListener("click", () => {
      renderTimeseriesTable(el, artifactId, offset + TS_TABLE_PAGE_SIZE);
    });
  } catch (err) {
    console.error("Table render failed:", err);
    el.innerHTML = '<span class="text-muted">Failed to load table data</span>';
  }
}

File: src/osprey/utils/timeseries.py
"""Timeseries utility functions.

Pure-math helpers for downsampling and extracting timeseries data,
used by both the Artifact Gallery interface and MCP server tools.
"""

import math


def _lttb_select_indices(values: list, max_points: int) -> list[int]:
    """Largest-Triangle-Three-Buckets index selection over one channel's values.

    Returns the indices to keep, never the values themselves: the caller slices
    its own ``timestamps``/``values`` with them, so a ``None`` gap marker
    (archiver disconnect / IOC reboot) is carried through untouched rather than
    being reconstructed. Nothing is manufactured -- every returned index refers
    to a sample that really exists.

    The x coordinate of a point is simply its index, so there is no separate
    x array: a point's position on the time axis is its ordinal, not its
    timestamp. (LTTB over unevenly spaced samples would weight by real elapsed
    time; this implementation deliberately does not, and never has.)

    Only reached with ``max_points >= 3`` and ``len(values) > max_points``;
    ``lttb_downsample_channel`` handles the pass-through cases before calling.

    Returns ``max_points`` ascending indices, always starting at 0 and ending
    at ``len(values) - 1``.
    """
    n = len(values)

    # Sanitized numeric working copy for triangle-area math ONLY.
    # None gap values -> 0.0 here so the bucket-average and area arithmetic
    # never see NoneType. Only this local copy is zero-filled; the caller
    # slices the ORIGINAL values, so gaps survive into the output.
    y = [0.0 if v is None else float(v) for v in values]

    selected = [0]  # Always keep first point
    bucket_size = (n - 2) / (max_points - 2)

    a_idx = 0
    for i in range(1, max_points - 1):
        # Next bucket boundaries
        b_start = int(math.floor((i - 1) * bucket_size)) + 1
        b_end = int(math.floor(i * bucket_size)) + 1
        b_end = min(b_end, n)

        # Average of the bucket after this one (lookahead)
        c_start = int(math.floor(i * bucket_size)) + 1
        c_end = int(math.floor((i + 1) * bucket_size)) + 1
        c_end = min(c_end, n)
        if c_start >= n:
            c_start = n - 1
        if c_end > n:
            c_end = n
        c_len = max(c_end - c_start, 1)
        # x[i] == i, so the mean of the look-ahead bucket's x coordinates is
        # just the mean of its index range -- no identity list materialized.
        avg_x = sum(range(c_start, c_end)) / c_len
        avg_y = sum(y[c_start:c_end]) / c_len

        # Pick point in current bucket with max triangle area
        max_area = -1.0
        best = b_start
        for j in range(b_start, b_end):
            area = abs((a_idx - avg_x) * (y[j] - y[a_idx]) - (a_idx - j) * (avg_y - y[a_idx]))
            if area > max_area:
                max_area = area
                best = j

        selected.append(best)
        a_idx = best

    selected.append(n - 1)  # Always keep last point
    return selected


def extract_channel_series(raw: dict) -> tuple[dict[str, dict], dict]:
    """Normalize any artifact timeseries layout into per-channel series.

    Handles three layouts:
      - New (long-format archiver_read): payload["series"] =
        {channel: {"timestamps": [...], "values": [...]}},
        payload["query"] = {...}. Already per-channel -- returned as-is.
      - Archiver (legacy split-orient): payload["dataframe"] =
        {columns, index, data}, payload["query"] = {...}.
      - Flat (legacy split-orient, no wrapper): payload =
        {columns, index, data} directly.

    For the two legacy split-orient layouts every column shares one `index`,
    so each column is transposed into its own (timestamps, values) pair.
    A `None` value at a given row/column means that channel had no sample at
    that shared timestamp -- not a real value -- so it is dropped rather than
    kept as a gap marker (contrast with ``lttb_downsample_channel``, which
    preserves a `None` gap that already survived into a channel's own column).

    Args:
        raw: Parsed JSON content of an artifact data file.

    Returns:
        Tuple of (series, query) where series maps channel name to
        {"timestamps": [...], "values": [...]}, in declaration order
        (dict/column order).
    """
    payload = raw.get("data", raw)

    if "series" in payload:
        return payload["series"], payload.get("query", {})

    if "dataframe" in payload:
        frame = payload["dataframe"]
        query = payload.get("query", {})
    else:
        frame = payload
        query = {}

    columns = frame.get("columns", [])
    index = frame.get("index", [])
    rows = frame.get("data", [])

    series: dict[str, dict] = {}
    for col_idx, channel in enumerate(columns):
        timestamps: list = []
        values: list = []
        for row_idx, row in enumerate(rows):
            value = row[col_idx] if col_idx < len(row) else None
            if value is None:
                continue
            timestamps.append(index[row_idx])
            values.append(value)
        series[channel] = {"timestamps": timestamps, "values": values}

    return series, query


def is_numeric_channel(values: list) -> bool:
    """True if every value in a channel's series is numeric (``None`` counts as numeric).

    This is the same check ``lttb_downsample_channel`` performs internally to decide
    whether LTTB's triangle-area math applies to a channel. Exposed so callers (the
    artifacts web API, ``archiver_downsample``) can report a channel's dtype directly
    rather than re-deriving it or guessing from a sample of values.

    Args:
        values: A channel's own values (numeric, ``None``, or enum/status strings).

    Returns:
        True if the channel is numeric (safe for LTTB), False otherwise.
    """
    # `None` counts as numeric: it is a gap marker (archiver disconnect / IOC
    # reboot), not an enum/status value, and LTTB zero-fills it for the area math
    # while preserving it in the output.
    return all(v is None or isinstance(v, int | float) for v in values)


def _even_subsample(timestamps: list, values: list, max_points: int) -> tuple[list, list]:
    """Evenly-spaced subsampling that always keeps the first and last points.

    Used in place of LTTB for non-numeric (enum/status) channels, where
    triangle-area math over the values is meaningless. Callers only reach
    this with ``max_points >= 3`` and ``len(timestamps) > max_points``.
    """
    # i=0 and i=max_points-1 land exactly on 0 and n-1, so first and last are
    # kept by construction. The set collapses the repeats rounding produces as
    # max_points approaches n; sorted restores order.
    n = len(timestamps)
    step = (n - 1) / (max_points - 1)
    selected = sorted({min(round(i * step), n - 1) for i in range(max_points)})
    return [timestamps[i] for i in selected], [values[i] for i in selected]


def lttb_downsample_channel(
    timestamps: list, values: list, max_points: int, numeric: bool | None = None
) -> tuple[list, list]:
    """Largest-Triangle-Three-Buckets downsampling for a single channel.

    Each channel carries its own timestamps, independent of every other
    channel's, so there is no shared x-axis and no multi-column slicing: this
    reduces exactly one (timestamps, values) pair. Callers with several
    channels call it once per channel.

    Selection and slicing are separate steps -- :func:`_lttb_select_indices`
    picks indices from a zero-filled numeric copy, and those indices are
    applied to the ORIGINAL ``timestamps``/``values``. That is what keeps a
    ``None`` gap a ``None`` in the output instead of a fabricated 0.0, and what
    guarantees the first and last samples are always returned.

    LTTB's triangle-area math requires numeric values. A non-numeric
    (enum/status) channel is never coerced toward that math -- it passes
    through unchanged when it already fits under ``max_points``, and is
    evenly subsampled (always keeping the first and last points) when it
    does not.

    A channel whose ``timestamps``/``values`` lengths differ (a malformed
    artifact, not the expected shape) is tolerated the same way
    ``_pivot_channel_series_to_table`` already tolerates it: both are truncated
    to the shorter length, so the mismatch never crashes the caller. Without
    this, real LTTB's helper arrays are built from ``values`` while its bucket
    loop walks ``range(len(timestamps))``, so a shorter ``values`` raises
    ``IndexError`` once the point count exceeds ``max_points``.

    Args:
        timestamps: This channel's own timestamps, ascending.
        values: This channel's own values, one per timestamp -- numeric,
            ``None``, or non-numeric strings (enum/status channels).
        max_points: Maximum number of points to return.
        numeric: Whether this channel is numeric, if the caller already knows.
            Callers that report a channel's dtype alongside its samples have
            already paid for :func:`is_numeric_channel`; passing the answer in
            keeps this from walking every value a second time. ``None`` means
            "work it out", which is what a caller that does not report dtype
            wants.

    Returns:
        Tuple of (downsampled_timestamps, downsampled_values).
    """
    if len(values) != len(timestamps):
        shared = min(len(timestamps), len(values))
        timestamps, values = timestamps[:shared], values[:shared]

    n = len(timestamps)
    if n <= max_points or max_points < 3:
        return timestamps, values

    if not (is_numeric_channel(values) if numeric is None else numeric):
        return _even_subsample(timestamps, values, max_points)

    selected = _lttb_select_indices(values, max_points)
    return [timestamps[i] for i in selected], [values[i] for i in selected]


def downsample_channel_map(
    series: dict[str, dict], max_points: int, channels: list[str] | None = None
) -> list[dict]:
    """Downsample each requested channel and report what that cost it.

    The shared body of every "give me a plottable version of this artifact"
    endpoint: pick the channels, reduce each one, and say how many points it
    started with, how many came back, and whether it is numeric. Callers rename
    these into their own published field names -- the artifacts web API and the
    ``archiver_downsample`` MCP tool each have their own wire vocabulary -- but
    none of them re-derive the numbers.

    Deciding ``numeric`` here rather than in each caller is also what keeps the
    channel from being walked twice: :func:`lttb_downsample_channel` needs the
    same answer internally, and takes it as an argument rather than recomputing
    it.

    Args:
        series: Per-channel series as returned by :func:`extract_channel_series`.
        max_points: Maximum points to return per channel.
        channels: Optional subset to keep, in the caller's requested order.
            ``None`` keeps every channel in the series' own order. Names not
            present in ``series`` are dropped, so an empty result means nothing
            requested matched.

    Returns:
        One record per selected channel, in selection order, with keys
        ``channel``, ``timestamps``, ``values``, ``original_points``,
        ``returned_points`` and ``numeric``.
    """
    selected = list(series) if channels is None else [c for c in channels if c in series]

    records = []
    for channel in selected:
        channel_series = series[channel]
        timestamps = channel_series.get("timestamps", [])
        values = channel_series.get("values", [])
        numeric = is_numeric_channel(values)
        ds_timestamps, ds_values = lttb_downsample_channel(
            timestamps, values, max_points, numeric=numeric
        )
        records.append(
            {
                "channel": channel,
                "timestamps": ds_timestamps,
                "values": ds_values,
                "original_points": len(timestamps),
                "returned_points": len(ds_timestamps),
                "numeric": numeric,
            }
        )
    return records
